feat: ignore-on-exit flag to control the error code

nirmo opened the pull request:
Fixes #155

Reference: [`ignore-on-exit`](https://docs.kics.io/latest/commands/) in Kics

Waits for #188

## Proposed Changes

<!--
- Assumed that the flag is relevant only for "plugins" sub-commands. Therefore added a Bool flag during the creation of the sub commands in the `main.go` file.

- The exit status for the mentioned scenarios is set in the `postRun` function hence added a bool variable to determine whether secrets should be ignored.
-->

- Ran manual tests with the confluence plugin:  
  | cmd | exit code |
  |--------|--------|
  | `2ms confluence --url <url> --spaces <space> --token <token> --username <>` | 1 |
  | `2ms confluence --url <url> --spaces <space> --token <token> --username <> --ignore-secrets` | 0 |

nirmo commented:
@baruchiro is the "--fail-on" a typo and should be "--ignore-on-exit"?

In addition, would you like just the name of the argument and the necessary behavioral changes to happen (from bool to string)? or should I look more in depth in the kics code for a different way?

Current implementation:
- Changed the argument according to Kics
- Used a function from Kics to validate the argument value (put into main.go)
- Set the argument validation in the preRun phase/func
- Changed the condition to return exit code 1 when secrets are found

Ran manual tests with the confluence plugin:
2ms confluence --url --spaces --token --username <> ; echo $? (result -> 1)
2ms confluence --url --spaces --token --username <> --ignore-on-exit results ; echo $? (result -> 0)
2ms confluence --url --spaces --token --username <> --ignore-on-exit ALL ; echo $? (result -> 0) // validation func will lower the value
2ms confluence --url --spaces --token --username <> --ignore-on-exit test ; echo $? (result -> 1)

baruchiro commented on cmd/main.go:
```diff
@@ -160,6 +164,11 @@ func validateFormat(stdout string, reportPath []string) {
 }
 
 func preRun(cmd *cobra.Command, args []string) {
+	shouldIgnore, _ := cmd.Flags().GetString("ignore-on-exit")
```
Even if you copied the code from Kics, please follow our code conventions.

You don't need to put the string literal, because you declared this string above.

And you don't even need to use `GetString` at all, because the value will already be in the `ignoreOnExitVar`.

baruchiro reviewed (changes requested):
I'm not sure I finished my review, but there are enough comments for now.

I think you copied the Kics code, but you didn't adjust it to our code.

Coping code is OK! I believe in this. But when I copy the code, I read it and learn it and make sure it is good at least as code I would write myself.

nirmo commented:
Background info about Kics workfolow:

console.excute (root command) returns an error

in case of a non-nil error, the ignore-on-exit value is checked with (ShowError func) in order to detemine whether a non zero exit call should be returned

One exception happenes when a scan results with a non-zero code. In that case, the program will exit using an explicit "os.Exit" call (post_scan.go) 


2ms Proposed Changes:

- Set an "error" return value to cmd.Execute() and made the relevant changes in the main fun to catch the error and print it.
- Added persistent flag "ignore-on-exit" to the rootCmd and not to the sub commands 
- Set an "error" return value to preRun and postRun commands
- Copied InitShouldIgnoreArg and ShowError fuctions from kics in order to validate the flag and detemine whether non-zero code should be returned
- Changed all log fatal calls but one (will discuss it later) to return an error (fmt.Errorf) instead.
- Made changes in the preRun command to catch an error from exisitng validateFormat and newly InitShouldIgnoreArg functions


There are at least two exceptions that I spotted:
1. Return zero code when secrets are found - in this case I followed Kics approach and returned an explicit "os.Exit" when needed.
2. Log fatal call in a go routine call in preRun func - I'm not with a way to bypass that so left it there. However I added an OR condition to an 
   existing if statement just before to check if errors should be printed and return "no error" if not. Not sure whether that's the right approach
   thought.

Further thoughts:
In kics there is a seperate package "helpers" to determine the exit code.

baruchiro commented on cmd/main.go:
```diff
@@ -205,17 +218,34 @@ func postRun(cmd *cobra.Command, args []string) {
 		if len(reportPathVar) > 0 {
 			err := report.WriteFile(reportPathVar, cfg)
 			if err != nil {
-				log.Error().Msgf("Failed to create report file with error: %s", err)
+				return fmt.Errorf("failed to create report file with error: %s", err)
 			}
 		}
 	} else {
-		log.Error().Msg("Scan completed with empty content")
-		os.Exit(0)
+		fmt.Println("Scan completed with empty content")
+		return nil
```
I think it will be better to finish the program always in one place, (on `main`).

To do that I can help you, I'm throwing here an idea from my mind, I didn't check it but I don't want to lose it.

1. Create an error channel.
2. From the go routine, send the error to the channel, or close the channel if no error.
3. After the go routine, read the error channel.
4. If you read an error, return it.
5. If you read `nil`, check the results count.

nirmo replied:
Sounds interesting, will have to do some research in order to better understand the concept and probably come back with some questions / code to discuss.

nirmo replied:
Proposed Changes:

log.fatal error in the preRun function:
- Added error channel
- Replaced log.fatal within the go routine under preRun with redirection of the err to the error channel 
- Added a check in the postRun function to determine whether the channel contains an error

Important Note:

While debugging I noticed that channels.WaitGroup.Wait() within postRun ends before the counter increase / first item is added.

Therefore it is basically skipped. Am I missing something?


TotalSecretsFound is greater than 0:

TotalSecretsFound is an exception case whether we want to exit with non-zero status code but avoid printing an error (as no error occurred).

As far as I could see, if error is returned from post/pre run then it will be printed along with a usage output (regardless of any intervention from our end).

As this is a special we could handle/bypass it by moving the actual "TotalSecretsFound > 0" condition to the Execute function after rootCmd.Execute().

In that case we will return an empty error to the main function. that would lead to a print of an empty line after the scan in case secrets are found.

A different idea is to not consider TotalSecretsFound as an error and add a check in the main function for that situation.


Further thoughts:
As errors are getting printed automatically from post/pre run functions. perhaps we can avoid the error print in the main function.

baruchiro reviewed (changes requested):
Thank you!

It was not as easy as I thought, sorry, but you're doing good work!

Maybe I have more comments, but I'm sending it now to not delay you.

nirmo commented on main.go:
```diff
@@ -20,7 +20,12 @@ func main() {
 	go listenForInterrupt(stopChan)
-	cmd.Execute()
+
+	if err := cmd.Execute(); err != nil {
+		if cmd.ShowError("errors") {
+			os.Exit(1)
+		}
+	}
```
I do see that errors are getting printed but can also add another print as I'm not sure when and how it is actually being done:
2ms filesystem --path /root/secret-test --ignore-on-exit allb; echo $?
Error: unknown argument for --ignore-on-exit: allb
valid arguments:
  none
  all
  results
  errors
Usage:
  2ms filesystem --path PATH [flags]

For the above erorr, it is printed after exiting/returning from the preRunE function but before being passed to the the main.go (err := cmd.Execute())

baruchiro replied:
Please format your message with [Markdown](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) to help me read it without trying to understand where is the code, where is the command and where is the error.

nirmo commented on cmd/main.go:
```diff
@@ -122,52 +123,62 @@ func Execute() {
 		subCommand.GroupID = group
-		subCommand.PreRun = preRun
-		subCommand.PostRun = postRun
+		subCommand.PreRunE = preRun
+		subCommand.PostRunE = postRun
```
yes. the error will be passed to:

if err := rootCmd.Execute(); err != nil {
		return err
	}

nirmo commented on cmd/main.go:
```diff
@@ -122,52 +123,62 @@ func Execute() {
+func validateFormat(stdout string, reportPath []string) error {
 	r := regexp.MustCompile(outputFormatRegexpPattern)
 	if !(r.MatchString(stdout)) {
-		log.Fatal().Msgf(`invalid output format: %s, available formats are: json, yaml and sarif`, stdout)
```
Sure, ran a quick grep and noticed less than 10 entries so will check them out and keep you updated.

baruchiro commented:
Hi, I tried to review your PR and I saw the `PreRun` is confusing and not written well, so I refactored it.
https://github.com/Checkmarx/2ms/pull/171

I know it will conflict your PR, but I think your work can be better and easier now.

Please let me know if you need help with the conflict or something else. Please do not get confused and re-create another PR to skip the conflict.

baruchiro commented on cmd/main.go:
```diff
@@ -197,31 +215,48 @@ func preRun(cmd *cobra.Command, args []string) {
+	return nil
+}
+
+// InitShouldIgnoreArg initializes what kind of errors should be used on exit codes
```
Remove comment

baruchiro commented on cmd/main.go:
```diff
@@ -197,31 +215,48 @@ func preRun(cmd *cobra.Command, args []string) {
+	return fmt.Errorf("unknown argument for --ignore-on-exit: %s\nvalid arguments:\n  %s", arg, strings.Join(validArgs, "\n  "))
+}
+
+// ShowError returns true if should show error, otherwise returns false
+func ShowError(kind string) bool {
+	return strings.EqualFold(ignoreOnExitVar, "none") || (!strings.EqualFold(ignoreOnExitVar, "all") && !strings.EqualFold(ignoreOnExitVar, kind))
 }
```
Remove the comment, and maybe split the long condition to make it readable

nirmo replied:
done

baruchiro commented on cmd/main.go:
```diff
@@ -130,43 +133,58 @@ func Execute() {
+		return err
+	}
+
+	if err := InitShouldIgnoreArg(ignoreOnExitVar); err != nil {
+		return err
 	}
```
I think you need to move it up as you can, since it defines the error behavior, it must be as first as we can, before other potential errors.

nirmo replied:
done

baruchiro commented on cmd/main.go:
```diff
@@ -77,6 +78,7 @@ var channels = plugins.Channels{
 
 var report = reporting.Init()
 var secretsChan = make(chan reporting.Secret)
+var errorChan = make(chan error, 1)
```
I think I don't see where you are using it.

nirmo replied:
Removed it

baruchiro commented on cmd/main.go:
```diff
@@ -197,31 +215,48 @@ func preRun(cmd *cobra.Command, args []string) {
 	// Show Report
 	if report.TotalItemsScanned > 0 {
-		report.ShowReport(stdoutFormatVar, cfg)
+		if err := report.ShowReport(stdoutFormatVar, cfg); err != nil {
+			return err
+		}
+
 		if len(reportPathVar) > 0 {
 			err := report.WriteFile(reportPathVar, cfg)
 			if err != nil {
-				log.Error().Msgf("Failed to create report file with error: %s", err)
+				return fmt.Errorf("failed to create report file with error: %s", err)
 			}
 		}
```
Where are you deciding if the exit code will be 1 or 0 for `TotalItemsScanned > 0`?

nirmo replied:
at the Execute() function

'''go
if err := rootCmd.Execute(); err != nil {
return err
}

if report.TotalSecretsFound > 0 && ShowError("errors") {
return fmt.Errorf("")
}
'''

baruchiro reviewed (commented):
Sorry for the delay

nirmo commented:
**Proposed Changes:**
- Simplified showError and added a check to verify whether err channel is empty,

**Next to do:**
- Remove log.Fatal messages from rule.go file

baruchiro commented on cmd/main.go:
```diff
@@ -261,7 +254,9 @@ func InitShouldIgnoreArg(arg string) error {
+		return true
+	}
+	return false
 }
```
I think you forgot to check the `kind`

nirmo replied:
Yes, you are right. I just fixed it but noticed another issue so will work on it later today/tomorrow and keep you updated.

nirmo replied:
**Proposed Changes:**
- Simplified and modified showError. Added a global variable (showErrorKind) to determine the behavior of the function.

**Next to do:**
- Remove log.Fatal messages from rule.go file

baruchiro commented:
@nirmo Please let me know when you're ready for the next review
